imbi/automations.py

```diff
@@ -10,66 +10,134 @@
+                self._add_error('GitLab project {} already exists for {}',
+                                result.row['gitlab_project_id'],
+                                result.row['project_slug'])
+
+        if not self._has_error():
+            return Project(
+                description=result.row['project_description'],
+                name=result.row['project_name'],
+                namespace=Namespace(
+                    gitlab_group_name=result.row['gitlab_group_name'],
+                    name=result.row['namespace_name'],
+                    slug=result.row['namespace_slug']),
+                project_id=project_id,
+                project_type=ProjectType(
+                    environment_urls=result.row['environment_urls'],
+                    gitlab_project_prefix=result.row['gitlab_project_prefix'],
+                    name=result.row['project_type_name'],
+                    slug=result.row['project_type_slug']),
+                slug=result.row['project_slug'])
+
+
+class GitLabCreateProjectAutomation(Automation):
+    def __init__(self,
+                 automation_settings: dict,
+                 project_id: int,
+                 user: 'imbi.user.User',
+                 db: sprockets_postgres.PostgresConnector):
+        super().__init__(user, db)
+        self.imbi_project_id = project_id
+        self.settings = automation_settings.get('gitlab', {})
         self._gitlab: typing.Optional[gitlab.GitLabClient] = None
         self._gitlab_parent: typing.Optional[dict] = None
+        self._project: typing.Optional[Project] = None
 
     async def prepare(self) -> typing.List[str]:
-        if self.imbi_project.gitlab_project_id is not None:
-            self._add_error('GitLab project {} already exists for {}',
-                            self.imbi_project.gitlab_project_id,
-                            self.imbi_project.slug)
-        else:
-            project_path, token = await asyncio.gather(
-                self._get_gitlab_project_path(), self._get_gitlab_token())
-            if project_path and token:
-                self._gitlab = gitlab.GitLabClient(token)
-                self._gitlab_parent = await self._get_gitlab_parent(
-                    project_path)
-
-        return self._errors
+        project: Project
+        token: imbi.integrations.IntegrationToken
+        project, token = await asyncio.gather(
+            self._get_project(self.imbi_project_id),
+            self._get_gitlab_token())
+        if project is not None and token is not None:
+            self._gitlab = gitlab.GitLabClient(token)
+            self._gitlab_parent = await self._get_gitlab_parent(project)
+            self._project = project
+        return self.errors
 
     async def run(self):
         gitlab_info = await self._gitlab.create_project(
-            self._gitlab_parent, self.imbi_project.name,
-            description=self.imbi_project.description)
+            self._gitlab_parent, self._project.name,
```

Comment:
You're referring to project as `self._project`. My understanding is that this means it is a protected variable, but I don't see where in the constructor it is being declared as private? I would expect `project_id: int` to look like `_project_id: int,`. Unless there is some Pythonic thing going on that I'm not aware of.

Reply:
The leading underscore means "please don't touch this outside of the implementation".  It is useful in class hierarchies though it might make sense to use a double-underscore here.  I use it to distinguish between the "public api" of the instance and things that you shouldn't be tinkering with.  For example `self.application`, `self.errors`, et.al. are attributes that you can use and depend on being consistent.  The underscore variables are used to cache information from the prepare phase to the run phase.  They aren't defined before `prepare` is called and shouldn't be used except from within the `run` method.  I should probably add some comments... maybe a decision register entry... about the implementation of automations.  The pattern ended up being a little more complex and twisted than I had expected.